### PR TITLE
[SPARK-34492][DOCS] Add "CSV Files" page for Data Source documents.

### PR DESCRIPTION
**Pull request by twoentartian**
### What changes were proposed in this pull request?

Fix [SPARK-34492], add Scala examples to read/write CSV files.  

### Why are the changes needed?

Fix [SPARK-34492].

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Build the document with "SKIP_API=1 bundle exec jekyll build", and everything looks fine.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @itholic

File: docs/sql-data-sources-csv.md
Comment (dongjoon-hyun) [inline docs/sql-data-sources-csv.md:54]: Hi, @twoentartian . Did you want to use `df3` instead of `df`?

File: docs/sql-data-sources-csv.md
Comment (HyukjinKwon) [inline docs/sql-data-sources-csv.md:29]: In addition, can we create examples files and link it here like we do in JSON? https://github.com/apache/spark/blob/7a4313ebf99d639c847a65361d32c1a031498cb0/docs/sql-data-sources-json.md

File: docs/sql-data-sources-csv.md
Comment (twoentartian) [reply docs/sql-data-sources-csv.md:54]: Using `df` seems better because of saving memory. But I think as an example, it doesn't matter.

File: docs/sql-data-sources-csv.md
Comment (twoentartian) [reply docs/sql-data-sources-csv.md:29]: good point, I haven't noticed it before. I will try to work on that. Should I close this pull request?

File: docs/sql-data-sources-csv.md
Comment (dongjoon-hyun) [reply docs/sql-data-sources-csv.md:54]: @twoentartian . What I mean is that this example is wrong. If you do `df.show`, you will get line 34 ~ 40.
> Using `df` seems better because of saving memory. But I think as an example, it doesn't matter.

Comment (twoentartian) [reply docs/sql-data-sources-csv.md:54]: Oh, I see that. I will correct this in the next pull request.

**Comment by HyukjinKwon** [comment]
@itholic would you like to pick the commits here and proceed further since the PR is dropped?

**Comment by itholic** [comment]
Sure, let me address.

**Comment by HyukjinKwon** [comment]
@twoentartian. are you still working on this? I assumed so because the PR was dropped (normally you can just push some more commits to this branch).

**Comment by twoentartian** [comment]
> @twoentartian. are you still working on this? I assumed so because the PR was dropped (normally you can just push some more commits to this branch).

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala
Comment (HyukjinKwon) [inline examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala:277]: Can you use `option` instead of `options(Map(...))` to be consistent with other examples?

**Comment by HyukjinKwon** [comment]
ok to test

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala
Comment (twoentartian) [reply examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala]: I add `.option()` and `.options()` to both Scala and Python examples.

**Comment by HyukjinKwon** [comment]
@itholic can you review this?

**Comment by itholic** [comment]
Sure, let me take a look

File: examples/src/main/python/sql/datasource.py
Comment (itholic) [inline examples/src/main/python/sql/datasource.py:279]: nit: Can we move this comment line to the above of the `df3.write.csv("output")` to keep consistency with other comments?

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala
Comment (itholic) [inline examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala:290]: ditto ?

**Review by itholic — COMMENTED** [review]
Anyway, shall we also have an examples for Java, R, or SQL ?

Otherwise, seems pretty fine to me.

**Review by srowen — COMMENTED** [review]
Seems OK - not sure there's much to know about reading CSV but docs are no bad thing.

File: examples/src/main/python/sql/datasource.py
Comment (twoentartian) [reply examples/src/main/python/sql/datasource.py:278]: Fixed.

File: examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala
Comment (twoentartian) [reply examples/src/main/scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala:289]: Fixed.

**Comment by twoentartian** [comment]
I can also write a Java page if necessary. Give me several days because I have some courses to do.
By the way, from this [page](https://sparkbyexamples.com/pyspark/pyspark-read-csv-file-into-dataframe/#Read%20all%20CSV%20files%20in%20a%20directory). It says we can read all csv files in a folder: `df = spark.read.csv("Folder path")`. But when I try it in spark-shell like this: 
`val path = "examples/src/main/resources"`
`val df3 = spark.read.option("delimiter", ";").option("header", "true").csv(path)`
The size of `df3` is 539 and seems it reads all files in that folder rather than csv file only. Is this API obsolete?

**Comment by srowen** [comment]
I think it's expected that it reads all files in the folder you specify?

**Comment by twoentartian** [comment]
> I think it's expected that it reads all files in the folder you specify?

image
This [page](https://sparkbyexamples.com/pyspark/pyspark-read-csv-file-into-dataframe/#Read%20all%20CSV%20files%20in%20a%20directory) says it should read all CSV files. Then it seems to be an error for this page.

**Comment by srowen** [comment]
I don't understand - it seems like it is doing what the documentation says then?

**Comment by twoentartian** [comment]
> I don't understand - it seems like it is doing what the documentation says then?

I think the page is indicating Spark will check whether it's a CSV file, if yes then read it, otherwise not. But in fact, Spark just reads all files without caring it's a CSV file or not.

**Comment by srowen** [comment]
Yes that's right, it is not paying attention to the suffix if you tell Spark that the files in the dir are CSV files. That's fine to add to the docs here, but is expected behavior IMHO.

**Comment by itholic** [comment]
Yeah, It is expected behavior but maybe we can describe more specifically to avoid such a confusion, such as "If path is specified as a directory, make sure that only CSV files exist in the directory because it reads all files other than CSV in the directory".

IMO it looks fine as it is now, but it would be better if a java page could be added.

But if you're busy, I can work with a follow-up if you want.

I'll leave it to the choice of you or other committers,

cc @HyukjinKwon @dongjoon-hyun and @srowen .

**Review by HyukjinKwon — APPROVED** [review]
Looks fine. Considering that it's a sub task of a bigger task in progress, and the author is first time contributor.

**Comment by twoentartian** [comment]
> Yeah, It is expected behavior but maybe we can describe more specifically to avoid such a confusion, such as "If path is specified as a directory, make sure that only CSV files exist in the directory because it reads all files other than CSV in the directory".
> 
> IMO it looks fine as it is now, but it would be better if a java page could be added.
> 
> But if you're busy, I can work with a follow-up if you want.
> 
> I'll leave it to the choice of you or other committers,
> 
> cc @HyukjinKwon @dongjoon-hyun and @srowen .

I will work on the Java document tomorrow. By the way, add a description about reading the directory.

**Comment by HyukjinKwon** [comment]
k, that would be great. It would be great if we can add SQL and R examples too to make it consistent with JSON file examples though. I am okay with separating it for this one if you're not used to R or SQL.

File: docs/sql-data-sources-csv.md
Comment (MaxGekk) [inline docs/sql-data-sources-csv.md:26]: I am not sure that we should refer to the outdated https://github.com/databricks/spark-csv . Does anybody maintain it? For example,
```
dateFormat: specifies a string that indicates the date format to use writing dates or timestamps. Custom date formats follow the formats at java.text.SimpleDateFormat. This applies to both DateType and TimestampType. If no dateFormat is specified, then "yyyy-MM-dd HH:mm:ss.S".
```
is incorrect.

Comment (HyukjinKwon) [reply docs/sql-data-sources-csv.md:26]: Yeah, we should fix it. Spark CSV is in Apache Spark and we should document it in more details :-).

Comment (twoentartian) [reply docs/sql-data-sources-csv.md]: I add a warning that some contents might be outdated in the [spark-csv](https://github.com/databricks/spark-csv).

**Comment by twoentartian** [comment]
> k, that would be great. It would be great if we can add SQL and R examples too to make it consistent with JSON file examples though. I am okay with separating it for this one if you're not used to R or SQL.

We have never touched R or SQL, so I am afraid we will not write further examples. We would like to write a summary here for `option()`, but we cannot locate where is the CSV source code. It would be great if you can help us.

File: docs/sql-data-sources-csv.md
Comment (srowen) [inline docs/sql-data-sources-csv.md:26]: Hm, rather than refer to spark-csv, which is outdated (because it is long since merged into Spark), we should inline some important docs from that site. I'd prefer to do some of that here, but, if you're short on time, we can leave this as a minimal placeholder for now.

File: docs/sql-data-sources-csv.md
Comment (srowen) [inline docs/sql-data-sources-csv.md:28]: Rather than call this out separately, maybe just say above "read a file or directory of files in CSV format"

File: docs/sql-data-sources-csv.md
Comment (srowen) [reply docs/sql-data-sources-csv.md:26]: Some edits:

save or write -> write
`option()` can be used to customize behavior of reading or writing, such as controlling behavior of the header, delimiter character, character set, and so on.

File: docs/sql-data-sources-csv.md
Comment (srowen) [inline docs/sql-data-sources-csv.md:36]: Don't duplicate this, I think? it's not language-specific. Keep it outside the language-specific blocks, once.

Comment (HyukjinKwon) [reply docs/sql-data-sources-csv.md:36]: Yeah, I think we don't need to mention about spark-csv here. Also, the options will be documented later via SPARK-34494

Comment (itholic) [reply docs/sql-data-sources-csv.md:36]: +1 for the comment from @srowen .

@twoentartian, How about write the contents simply such as

"Spark SQL support read and write methods for CSV file, to read a CSV file into Spark DataFrame and to write the Spark DataFrame to the CSV file. `option()` can be used to customize behavior of reading or writing, such as controlling behavior of the header, delimiter character, character set, and so on."

outside of the language-specific block, after then maybe we can just simply show the code example for each language.

Refer to the "Hive Tables" page for the reference.

Screen Shot 2021-03-30 at 6 02 29 PM

**Comment by itholic** [comment]
LGTM except the https://github.com/apache/spark/pull/31827#discussion_r603915655.

File: docs/sql-data-sources-csv.md
Comment (twoentartian) [reply docs/sql-data-sources-csv.md]: Fixed.

File: docs/sql-data-sources-csv.md
Comment (twoentartian) [reply docs/sql-data-sources-csv.md]: Fixed.

Comment (twoentartian) [reply docs/sql-data-sources-csv.md]: > Hm, rather than refer to spark-csv, which is outdated (because it is long since merged into Spark), we should inline some important docs from that site. I'd prefer to do some of that here, but, if you're short on time, we can leave this as a minimal placeholder for now.

I leave a comment (`<!--TODO: add `option()` document reference--> `) in the markdown file as the placeholder for `option()`.

File: docs/sql-data-sources-csv.md
Comment (twoentartian) [reply docs/sql-data-sources-csv.md]: I changed it to this :
image

Comment (twoentartian) [reply docs/sql-data-sources-csv.md]: > Yeah, I think we don't need to mention about spark-csv here. Also, the options will be documented later via [SPARK-34494](https://issues.apache.org/jira/browse/SPARK-34494)

In addition, it seems the `option()` will be documented separately in the future, so I leave a comment in the markdown file like this:
image

**Review by srowen — APPROVED** [review]

**Comment by itholic** [comment]
LGTM.

Thanks, @twoentartian .

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
@twoentartian do you mind adding a comment in the JIRA so I can identify your JIRA id and assign SPARK-34492 to you?

**Pull request closed without merging**
